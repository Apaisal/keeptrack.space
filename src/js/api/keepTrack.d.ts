/* eslint-disable no-unused-vars */
export declare interface SatObject {
  id: number;
  satId?: number;
  position: { x: number; y: number; z: number };
  static?: boolean;
  missile?: boolean;
  type?: string;
  name?: string;
  intlDes?: string;
  TLE1: string;
  TLE2: string;
  SCC_NUM: string;
  active: boolean;
  C?: string;
  LS?: string;
  LV?: string;
  ON: string;
  OT: number;
  TLE1: string;
  TLE2: string;
  R: string;
  URL?: string;
  O?: string;
  U?: string;
  P?: string;
  LM?: string;
  DM?: string;
  Pw?: string;
  Li?: string | number;
  Con?: string;
  M?: string;
  S1?: string;
  S2?: string;
  S3?: string;
  S4?: string;
  S5?: string;
  S6?: string;
  S7?: string;
  inclination: number;
  lon: number;
  perigee: number;
  apogee: number;
  period: number;
  meanMotion: number;
  semimajorAxis: number;
  eccentricity: number;
  raan: number;
  argPe: number;
  inView: number;
  velocity: {
    total: number;
    x: number;
    y: number;
    z: number;
  };
  getTEARR?: any;
  getAltitude?: any;
  getDirection?: any;
  vmag?: number;
  associates?: any;
  maneuver?: any;
  constellation?: any;
  ORPO?: any;
  FMISSED?: any;
  NOTES?: any;
  TTP?: any;
  isInSun?: any;
  inSun?: any;
  desc?: string;
  marker?: boolean;
  isRadarData?: boolean;
  isInGroup?: boolean;
  missileComplex?: number;
}

export interface SensorObject {
  obsmaxel2?: any;
  obsmaxrange2?: any;
  obsminrange2?: any;
  obsminel2?: any;
  obsmaxaz2?: any;
  obsminaz2?: any;
  alt: number;
  beamwidth?: number;
  changeObjectInterval?: number;
  country: string;
  lat: number;
  linkAehf?: boolean;
  linkWgs?: boolean;
  lon: number;
  name: string;
  alt: number;
  observerGd: {
    lat: number;
    lon: number;
    alt: number;
  };
  obsmaxaz: number;
  obsmaxel: number;
  obsmaxrange: number;
  obsminaz: number;
  obsminel: number;
  obsminrange: number;
  shortName: string;
  staticNum: number;
  sun: string;
  type?: string;
  url?: string;
  volume: boolean;
  zoom: string;
}

export interface Lla {
  lat: number;
  lon: number;
  alt: number;
}

export interface Rae {
  rng: number;
  az: number;
  el: number;
}

export interface InView {
  inView: boolean;
}

export interface EarthObject {
  init: any;
  draw: any;
  program: any;
  update: any;
  drawOcclusion: any;
  earthJ: number;
  earthEra: number;
  lightDirection: any;
  pos: [number, number, number];
  loaded: boolean;
  sunvar: any;
  shader: any;
  specularMap: any;
  imgHiRes: any;
  isHiResReady: boolean;
  nightImgHiRes: any;
  nightImg: any;
  bumpMap: any;
  loadHiResNight: any;
  loadHiRes: any;
  isUseHiRes: boolean;
  lightDirection: [number, number, number];
}

export interface keepTrackApiInterface {
  html: (strings: TemplateStringsArray, ...placeholders: any[]) => string;
  register: (params: { method: string; cbName: string; cb: any }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}
