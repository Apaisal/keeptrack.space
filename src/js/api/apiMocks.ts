import { satObject, sensorObject } from './keepTrack';

export const defaultSat: satObject = {
  active: true,
  SCC_NUM: '00005',
  C: 'ISS',
  LS: 'TTMTR',
  LV: 'Proton-K',
  ON: 'ISS (ZARYA)',
  OT: 1,
  TLE1: '1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991',
  TLE2: '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053',
  R: '99.0524',
  URL: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  O: 'National Aeronautics and Space Administration (NASA)/Multinational',
  U: 'Government',
  P: 'Space Science',
  LM: '',
  DM: '',
  Pw: '',
  Li: 30,
  Con: 'Boeing Satellite Systems (prime)/Multinational',
  M: 'Final size of a Boeing 747; first component.',
  S1: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  S2: '',
  S3: 'http://www.boeing.com/defense-space/space/spacestation/overview/',
  S4: '',
  S5: '',
  S6: '',
  S7: '',
  inclination: 51.6423,
  longitude: 168.5744,
  perigee: 1475,
  apogee: 184.3976,
  period: 313.3642,
  meanMotion: 15.48839820294053,
  semimajorAxis: 21203.40407588,
  eccentricity: 0.00003453,
  raan: 0,
  argPe: 0,
  inview: 1,
  velocity: {
    total: 0,
    x: 0,
    y: 0,
    z: 0,
  },
  getTEARR: () => ({
    lon: 168.5744,
    lat: 51.6423,
    alt: 1475,
  }),
  getAltitude: () => 0,
  getDirection: () => 'N',
};

export const defaultSensor: sensorObject = {
  alt: 0.060966,
  beamwidth: 2,
  changeObjectInterval: 1000,
  country: 'United States',
  lat: 41.754785,
  linkAehf: true,
  linkWgs: true,
  lon: -70.539151,
  name: 'Cape Cod AFS, Massachusetts',
  observerGd: { lat: 0.7287584767123405, lon: -1.2311404365114507, alt: 0.060966 },
  obsmaxaz: 227,
  obsmaxel: 85,
  obsmaxrange: 5556,
  obsminaz: 347,
  obsminel: 3,
  obsminrange: 200,
  shortName: 'COD',
  staticNum: 0,
  sun: 'No Impact',
  type: 'Phased Array Radar',
  url: 'http://www.radartutorial.eu/19.kartei/01.oth/karte004.en.html',
  volume: false,
  zoom: 'leo',
};

export const useMockWorkers = () => {
  class Worker {
    url: any;
    onmessage: (msg: any) => void;
    constructor(stringUrl: any) {
      this.url = stringUrl;
      this.onmessage = () => {
        return {
          data: {}
        }
      };
    }

    postMessage(msg: any) {
      if (msg.dat) {
        this.onmessage({          
          data: {
            extraData: JSON.stringify([defaultSat, defaultSensor]),
            satPos: [0,0,0,0,0,0],
            satVel: [0,0,0,0,0,0],
          }
        });
      }
      this.onmessage({
        data: {
          satPos: [0,0,0,0,0,0],
          satVel: [0,0,0,0,0,0],
        }
      });
    }
  }

  (<any>window).Worker = Worker;
}

export const keepTrackApiStubs = {
  programs: {
    adviceManager: {
      adviceList: {
        satViewDisabled: jest.fn(),
        editSatDisabled: jest.fn(),
        survFenceDisabled: jest.fn(),
        bubbleDisabled: jest.fn(),
        satFOVDisabled: jest.fn(),
        planetariumDisabled: jest.fn(),
        mapDisabled: jest.fn(),
      },
      adviceCount: {
        socrates: 0,
      },
      showAdvice: jest.fn(),
      adviceArray: (<any[]>[]),
    },
    mainCamera: {
      cameraType: {
        current: 1,
        satellite: 1,
        fixedToSat: 2,
      },
      camSnap: jest.fn(),
      changeZoom: jest.fn(),
      latToPitch: jest.fn(),
      longToYaw: jest.fn(),
      autoRotate: jest.fn(),
    },
    ColorScheme: {
      reloadColors: jest.fn(),      
    },
    drawManager: {
      glInit: jest.fn(),
      gl: global.mocks.glMock,
      selectSatManager: {
        selectSat: jest.fn(),
      },
    },    
    dotsManager: {
      starIndex1: 0,
      pickReadPixelBuffer: [
        0,
        0,
        0,
      ],
    },    
    groups: null,
    groupsManager: {
      Canada: null,
      Japan: null,
      France: null,
      China: null,
      India: null,
      Israel: null,
      Russia: null,
      UnitedKingdom: null,
      UnitedStates: null,
      SpaceStations: {
        sats: [
          {satId: 1, SCC_NUM: '25544'},
          {satId: 1, SCC_NUM: '25544'},
        ],
        updateOrbits: jest.fn(),
      },
      clearSelect: jest.fn(),
      createGroup: () => ({
        updateOrbits: jest.fn(),
      }),
      selectGroup: jest.fn(),
    },    
    lineManager: {
      create: jest.fn(),
      removeStars: jest.fn(),
      getLineListLen: jest.fn(),
    },
    missileManager: {
      getMissileTEARR: jest.fn(),
    },
    objectManager: {
      isLaunchSiteManagerLoaded: true,
      launchSiteManager: {
        launchSiteList: [
          'AFWTR'
        ],
      },
      selectedSat: -1,
      setSelectedSat: jest.fn(),
      staticSet: {
        id: 0,
      },
      analSatSet: {
        id: 0,
      },
      radarDataSet: {
        id: 0,
      },
      missileSet: {
        id: 0,
      },
      fieldOfViewSet: {
        id: 0,
      },
    },
    orbitManager: {
      updateOrbitBuffer: jest.fn(),
      clearInViewOrbit: jest.fn(),
      removeInViewOrbit: jest.fn(),
      historyOfSatellitesPlay: jest.fn(),
    },
    satellite: {
      currentTEARR: {
        az: 0,
        el: 0,
        rng: 0,
      },
      sgp4: () => ([0,0,0]),
      setobs: jest.fn(),
      setTEARR: jest.fn(),
      altitudeCheck: jest.fn(),
      degreesLong: (num: number) => num,
      degreesLat: (num: number) => num,
      currentEpoch: () => ['21', '150'],
      getOrbitByLatLon: () => ['1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991', '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053'],
      nextpassList: () => [new Date().getTime(), new Date().getTime() + 1],
      twoline2satrec: () => ({ jdsatepoch: 2458000.5 }),
      getRae: () => [0, 0, 0],
      checkIsInFOV: () => true,
      findCloseObjects: () => '',
      findBestPasses: () => jest.fn(),
      nextpass: () => '2021-09-21 20:00:00Z',
      getDOPsTable: jest.fn(),
    },
    satSet: {
      satCruncher: {
        postMessage: jest.fn(),
      },
      satData: [
        defaultSat,
        defaultSat,
      ],
      getIdFromObjNum: jest.fn(),
      getSensorFromSensorName: () => 0,
      getSatExtraOnly: () => defaultSat,
      getSatFromObjNum : () => defaultSat,
      getSat: () => defaultSat,
      initGsData: jest.fn(),
      setColorScheme: jest.fn(),
      getSatInViewOnly: () => defaultSat,
      default: '',
      insertNewAnalystSatellite: jest.fn(),
    },
    searchBox: {
      hideResults: jest.fn(),
      fillResultBox: jest.fn(),
    },
    sensorFov: {
      enableFovView: jest.fn(),
    },
    sensorManager: {
      checkSensorSelected: () => false,
      currentSensor: {
        alt: 0,
        lat: 0,
        lon: 0,
        shortName: 'COD',
      },
    },
    starManager: {
      init: jest.fn(),
    },
    timeManager: {
      propOffset: 0,
      propRate: 0,
      selectedDate: 0,
      propTime: () => 0,
      getDayOfYear: () => 0,
    },
    uiManager: {
      hideSideMenus: jest.fn(),
      toast: jest.fn(),
      colorSchemeChangeAlert: jest.fn(),
      bottomIconPress: jest.fn(),
      legendColorsChange: jest.fn(),
      doSearch: jest.fn(),
      loadStr: jest.fn(),
      searchToggle: jest.fn(),
      legendMenuChange: jest.fn(),
      clearRMBSubMenu: jest.fn(),
      updateURL: jest.fn(),
    },
    watchlist: {
      updateWatchlist: jest.fn(),
    },
  },
};

keepTrackApiStubs.programs.groupsManager.Canada = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Japan = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.France = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.China = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.India = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Israel = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Russia = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedKingdom = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedStates = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groups = keepTrackApiStubs.programs.groupsManager;