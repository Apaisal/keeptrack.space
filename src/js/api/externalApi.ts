interface keepTrackApiInterface {
  register: (params: { method: string; cbName: string; cb: () => void }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}

const keepTrackApi: keepTrackApiInterface = {
  register: (params: { method: string; cbName: string; cb: () => void }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      // Add the callback
      keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
    } else {
      throw new Error(`Invalid callback "${params.method}"!`);
    }
    return;
  },
  unregister: (params: { method: string; cbName: string }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
        if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
          keepTrackApi.callbacks[params.method].splice(i, 1);
          return;
        }
      }
      // If we got this far, it means we couldn't find the callback
      throw new Error(`Callback "${params.cbName} not found"!`);
    } else {
      // Couldn't find the method
      throw new Error(`Invalid callback "${params.method}"!`);
    }
  },
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      for (let i = 0; i < keepTrackApi.callbacks.selectSatData.length; i++) {
        keepTrackApi.callbacks.selectSatData[i].cb(sat, satId);
      }
    },
    updateSelectBox: (sat: any) => {
      for (let i = 0; i < keepTrackApi.callbacks.updateSelectBox.length; i++) {
        keepTrackApi.callbacks.updateSelectBox[i].cb(sat);
      }
    },
  },
  programs: {
    timeManager: {},
    settingsManager: {},
    ColorScheme: {},
    drawManager: {},
    missileManager: {},
    objectManager: {},
    orbitManager: {},
    photoManager: {},
    satSet: {},
    satellite: {},
    searchBox: {},
    sensorManager: {},
    starManager: {},
    sMM: {},
    uiManager: {},
    uiInput: {},
  },
};

(<any>window).keepTrackApi = keepTrackApi;
export { keepTrackApi };
