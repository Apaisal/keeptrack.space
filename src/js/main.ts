/* eslint-disable no-unreachable */
/**
 * /*! /////////////////////////////////////////////////////////////////////////////
 *
 * main.js is the primary javascript file for keeptrack.space. It manages all user
 * interaction with the application.
 * http://keeptrack.space
 *
 * @Copyright (C) 2016-2022 Theodore Kruczek
 * @Copyright (C) 2020 Heather Kruczek
 * @Copyright (C) 2015-2016, James Yoder
 *
 * Original source code released by James Yoder at https://github.com/jeyoder/ThingsInSpace/
 * under the MIT License. Please reference http://keeptrack.space/license/thingsinspace.txt
 *
 * KeepTrack is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * KeepTrack is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with
 * KeepTrack. If not, see <http://www.gnu.org/licenses/>.
 *
 * /////////////////////////////////////////////////////////////////////////////
 */

import { isThisJest, keepTrackApi } from './api/keepTrackApi';
import { importCss } from './css';

export const showErrorCode = (error: Error & { lineNumber: number }): void => {
  let errorHtml = '';
  errorHtml += error?.message ? `${error.message}<br>` : '';
  errorHtml += error?.lineNumber ? `Line: ${error.lineNumber}<br>` : '';
  errorHtml += error?.stack ? `${error.stack}<br>` : '';
  document.getElementById('loader-text').innerHTML = errorHtml;
  // istanbul ignore next
  if (!isThisJest()) console.warn(error);
};

export const loadAfterStart = () => {
  if (settingsManager.cruncherReady) {
    // Update any CSS now that we know what is loaded
    keepTrackApi.methods.uiManagerFinal();
    // Update MaterialUI with new menu options
    window.M.AutoInit();
  } else {
    setTimeout(loadAfterStart, 100);
  }
};

// Load the CSS
importCss();
// Load the main website
import('./initalizeKeepTrack').then(({ initalizeKeepTrack }) => {
  initalizeKeepTrack();
});
