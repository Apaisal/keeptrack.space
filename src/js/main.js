/**
 * /*! /////////////////////////////////////////////////////////////////////////////
 *
 * main.js is the primary javascript file for keeptrack.space. It manages all user
 * interaction with the application.
 * http://keeptrack.space
 *
 * @Copyright (C) 2016-2021 Theodore Kruczek
 * @Copyright (C) 2020 Heather Kruczek
 * @Copyright (C) 2015-2016, James Yoder
 *
 * Original source code released by James Yoder at https://github.com/jeyoder/ThingsInSpace/
 * under the MIT License. Please reference http://keeptrack.space/license/thingsinspace.txt
 *
 * KeepTrack is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * KeepTrack is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with
 * KeepTrack. If not, see <http://www.gnu.org/licenses/>.
 *
 * /////////////////////////////////////////////////////////////////////////////
 */

import '@app/js/api/externalApi';
import 'jquery-ui-bundle';
import 'materialize-css';
import { uiInput, uiManager } from '@app/js/uiManager/uiManager.js';
import { Camera } from '@app/js/cameraManager/camera.js';
import { ColorSchemeFactory as ColorScheme } from '@app/js/colorManager/color-scheme-factory.js';
import { GroupFactory } from '@app/js/groupsManager/group-factory.js';
import { LineFactory } from '@app/js/drawManager/sceneManager/sceneManager.js';
import { drawManager } from '@app/js/drawManager/drawManager.js';
import { jQAlt } from '@app/js/lib/jqalt.js';
import { missileManager } from '@app/js/missileManager/missileManager.js';
import { objectManager } from '@app/js/objectManager/objectManager.js';
import { orbitManager } from '@app/js/orbitManager/orbitManager.js';
import { photoManager } from '@app/js/photoManager/photoManager.js';
// import { radarDataManager } from'@app/js/satSet/radarDataManager.js';
import { sMM } from '@app/js/uiManager/sideMenuManager.js';
import { satSet } from '@app/js/satSet/satSet.js';
import { satellite } from '@app/js/lib/lookangles.js';
import { searchBox } from '@app/js/uiManager/search-box.js';
import { sensorManager } from '@app/js/sensorManager/sensorManager.js';
import { settingsManager } from '@app/js/settingsManager/settingsManager.ts';
import { starManager } from '@app/js/starManager/starManager.js';
import { timeManager } from '@app/js/timeManager/timeManager.js';

const keepTrackApi = window.keepTrackApi;
keepTrackApi.programs.ColorScheme = ColorScheme;
keepTrackApi.programs.drawManager = drawManager;
keepTrackApi.programs.missileManager = missileManager;
keepTrackApi.programs.objectManager = objectManager;
keepTrackApi.programs.orbitManager = orbitManager;
keepTrackApi.programs.photoManager = photoManager;
keepTrackApi.programs.satSet = satSet;
keepTrackApi.programs.satellite = satellite;
keepTrackApi.programs.searchBox = searchBox;
keepTrackApi.programs.sensorManager = sensorManager;
keepTrackApi.programs.settingsManager = settingsManager;
keepTrackApi.programs.starManager = starManager;
keepTrackApi.programs.sMM = sMM;
keepTrackApi.programs.timeManager = timeManager;
keepTrackApi.programs.uiManager = uiManager;
keepTrackApi.programs.uiInput = uiInput;

const initalizeKeepTrack = async () => {
  try {
    await timeManager.init();
    settingsManager.loadStr('dots');
    uiManager.mobileManager.init();
    const cameraManager = new Camera();
    keepTrackApi.programs.cameraManager = cameraManager;
    // We need to know if we are on a small screen before starting webgl
    await drawManager.glInit();
    window.addEventListener('resize', drawManager.resizeCanvas);

    drawManager.loadScene();

    const dotsManager = await drawManager.createDotsManager();
    keepTrackApi.programs.dotsManager = dotsManager;

    satSet.init();
    objectManager.init();
    await ColorScheme.init();
    drawManager.selectSatManager.init();

    await satSet.loadCatalog(); // Needs Object Manager and gl first
    const satCruncher = satSet.satCruncher;
    keepTrackApi.programs.satCruncher = satCruncher;

    dotsManager.setupPickingBuffer(satSet.satData);
    satSet.setColorScheme(ColorScheme.default, true);

    const groupsManager = new GroupFactory();
    keepTrackApi.programs.groupsManager = groupsManager;

    await orbitManager.init();
    searchBox.init();

    const lineManager = new LineFactory();
    keepTrackApi.programs.lineManager = lineManager;

    starManager.init();
    uiManager.init();
    await satellite.initLookangles();
    dotsManager.updateSizeBuffer(satSet.satData);
    // await radarDataManager.init(sensorManager, satSet, satCruncher, satellite);
    satSet.setColorScheme(settingsManager.currentColorScheme); // force color recalc
    objectManager.satLinkManager.idToSatnum(satSet);

    uiInput.init();

    drawManager.init();

    // Now that everything is loaded, start rendering to thg canvas
    drawManager.drawLoop();

    // UI Changes after everything starts -- DO NOT RUN THIS EARLY IT HIDES THE CANVAS
    uiManager.postStart();
    photoManager.init();
  } catch (error) {
    /* istanbul ignore next */
    console.warn(error);
  }
};

jQAlt.docReady(initalizeKeepTrack);

// For testing purposes
export { initalizeKeepTrack };
