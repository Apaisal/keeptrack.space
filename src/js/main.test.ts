import { keepTrackApiStubs } from './api/apiMocks';
import { keepTrackApi } from './api/keepTrackApi';
import * as main from './main';

keepTrackApi.programs = <any>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs };

const setUrl = (url) => {
  const host = url.split('/')[2] || '';
  let search = url.split('?')[1] || '';
  search = search !== '' ? `?${search}` : '';

  global.window = Object.create(window);
  Object.defineProperty(window, 'location', {
    value: {
      href: url,
      host: host,
      search: search,
    },
    writable: true,
  });
};

// @ponicode
describe('main.importCss', () => {
  test('0', async () => {
    const result = await main.importCss();
    expect(result).toMatchSnapshot();
  });
});

describe('main.forceHttps', () => {
  it('should change an http to https', () => {
    setUrl('http://dummy.com');
    main.forceHttps();
    expect(window.location.href).toBe('https://dummy.com');
  });
  it('should not break an https request', () => {
    setUrl('https://dummy.com');
    main.forceHttps();
    expect(window.location.href).toBe('https://dummy.com');
  });
});

describe('main.showErrorCode', () => {
  it('should show the error code', () => {
    document.body.innerHTML = '<div id="loader-text"></div>';
    const element = document.getElementById('loader-text');
    main.showErrorCode(<Error & { lineNumber: number }>(<unknown>{ message: 'test', lineNumber: 1, stack: 'test' }));
    expect(element.innerHTML).toMatchSnapshot();
  });
});

// @ponicode
describe('main.initalizeKeepTrack', () => {
  test('0', async () => {
    jest.setTimeout(60 * 1000);
    keepTrackApi.methods.loadCatalog = jest.fn();
    await main.initalizeKeepTrack();
  });

  it('should be a function', () => {
    expect(main.initalizeKeepTrack).toBeInstanceOf(Function);
  });

  it('should return a promise', () => {
    expect(main.initalizeKeepTrack()).toBeInstanceOf(Promise);
  });

  it('should not throw any errors', async () => {
    const spy = jest.spyOn(main, 'showErrorCode');
    const result = await main.initalizeKeepTrack();
    expect(() => result).not.toThrow();
    expect(spy).not.toBeCalled();
  });
});
