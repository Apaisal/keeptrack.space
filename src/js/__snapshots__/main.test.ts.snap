// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main.redirectHttpToHttps 0 1`] = `undefined`;

exports[`main.showErrorCode 0 1`] = `undefined`;

exports[`main.showErrorCode 1 1`] = `undefined`;

exports[`main.showErrorCode 2 1`] = `undefined`;

exports[`main.showErrorCode 3 1`] = `undefined`;

exports[`main.showErrorCode 4 1`] = `undefined`;

exports[`main.showErrorCode 5 1`] = `undefined`;
